### PR TITLE
Move is_complex() directly onto Tensor

### PR DESCRIPTION
**Pull request by jbschlosser**
Addresses `is_complex` for #44809

Removes `is_complex` from `native_functions.yaml` and implements it directly within `TensorBody.h` / `TensorMethods.cpp`. Supports both function and method forms and exposes them to Python.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2082,10 +2082,6 @@
   variants: function, method
   device_guard: False
 
-- func: is_complex(Tensor self) -> bool
-  variants: function, method
-  device_guard: False
-
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:2088]: The error logs are suggesting that either the type hint for `is_complex` is missing, or the actual `Tensor.is_complex` method is missing.

I suspect the problem is the former. @jbschlosser could you check if `is_complex` is in `torch/_C/__init__.pyi` (this is an autogenerated file) on your local branch? If not we have to manually add that type hint somewhere
```
Feb 10 19:33:01 ======================================================================
Feb 10 19:33:01 FAIL [69.756s]: test_run_mypy (__main__.TestTypeHints) [mypy.ini]
Feb 10 19:33:01 ----------------------------------------------------------------------
Feb 10 19:33:01 Traceback (most recent call last):
Feb 10 19:33:01   File "test_type_hints.py", line 171, in test_run_mypy
Feb 10 19:33:01     self.fail(f"mypy failed: {stdout} {stderr}")
Feb 10 19:33:01 AssertionError: mypy failed: torch/testing/__init__.py:108: error: "Tensor" has no attribute "is_complex"  [attr-defined]
Feb 10 19:33:01 torch/testing/_internal/common_nn.py:5106: error: "Tensor" has no attribute "is_complex"  [attr-defined]
Feb 10 19:33:01 torch/testing/_internal/common_methods_invocations.py:691: error: "Tensor" has no attribute "is_complex"  [attr-defined]
Feb 10 19:33:01 torch/testing/_internal/common_methods_invocations.py:3308: error: "Tensor" has no attribute "is_complex"  [attr-defined]
Feb 10 19:33:01 Found 4 errors in 3 files (checked 1229 source files)
```

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -1027,7 +1027,6 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   m.impl("_remove_batch_dim", native::_remove_batch_dim);
 
   m.impl("sum.dim_IntList", sum_batching_rule);
-  m.impl("is_complex", native::is_complex);
```
Comment (zou3519) [inline aten/src/ATen/BatchingRegistrations.cpp:1030]: Change to BatchingRegistrations lgtm as long as the tests are passing

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:2088]: Nice, thanks for catching this! I updated `tools/pyi/gen_pyi.py` to add the entry for `is_complex` and the type hint tests pass now.

**Review by glaringlee — COMMENTED** [review]
Seems `is_complex` is not jit script-able after this change, will dig this more.

**Comment by zou3519** [comment]
> Seems `is_complex` is not jit script-able after this change, will dig this more.

Originally the JIT codegen generates some logic to handle `is_complex` from native_functions.yaml. Now that it is not in native_functions.yaml, we need to add it manually. You can follow the example of `is_contiguous` here:
https://github.com/pytorch/pytorch/blob/7763c127cd5630ba4123ad89fc5243c28e91aa4a/torch/csrc/jit/runtime/register_prim_ops.cpp#L535-L545

**Comment by jbschlosser** [comment]
Closing because it is no longer necessary due to the previously merged PR #50054.

**Pull request closed without merging**